File: src/backend/Domain/Models/Spool.cs
namespace Domain.Models;

public class Spool
{
    public Guid Id { get; set; }
    public string Brand { get; set; } = string.Empty;
    public string Material { get; set; } = string.Empty;
    public string ColorName { get; set; } = string.Empty;
    public string ColorHex { get; set; } = string.Empty;
    public float InitialWeightG { get; set; }
    public float CurrentWeightG { get; set; }
    public float SpoolWeightG { get; set; } = 200;
    public float LowStockThresholdG { get; set; } = 100;
    public bool IsActive { get; set; } = false;
    public bool IsArchived { get; set; } = false;
    public DateTime? ArchivedAt { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime? LastScannedAt { get; set; }
    public string? Notes { get; set; }
    public string? StockLocation { get; set; }
    public decimal? Price { get; set; }
    public float? Density { get; set; }
    public float? DiameterTolerance { get; set; }
    public int? ExtruderMin { get; set; }
    public int? ExtruderMax { get; set; }
    public int? BedMin { get; set; }
    public int? BedMax { get; set; }
    public ICollection<NfcTag> NfcTags { get; set; } = new List<NfcTag>();
    public ICollection<PrintJob> PrintJobs { get; set; } = new List<PrintJob>();
}

File: src/frontend/src/types/spool.ts
export interface SpoolResponse {
  id: string
  brand: string
  material: string
  colorName: string
  colorHex: string
  initialWeightG: number
  currentWeightG: number
  spoolWeightG: number
  lowStockThresholdG: number
  isActive: boolean
  isArchived: boolean
  createdAt: string
  lastScannedAt: string | null
  notes: string | null
  density: number | null
  diameterTolerance: number | null
  extruderMin: number | null
  extruderMax: number | null
  bedMin: number | null
  bedMax: number | null
  hasNfcTag: boolean
  nfcTagUid: string | null
  nfcTagUids?: string[]
  printerId: string | null
  printerName: string | null
  amsSlot: number | null
  price: number | null
  stockLocation: string | null
}

export interface UpdateSpoolRequest {
  brand?: string
  material?: string
  colorName?: string
  colorHex?: string
  currentWeightG?: number
  initialWeightG?: number
  spoolWeightG?: number
  lowStockThresholdG?: number
  notes?: string
  isActive?: boolean
  price?: number | null
  stockLocation?: string
  density?: number
  diameterTolerance?: number
  extruderMin?: number
  extruderMax?: number
  bedMin?: number
  bedMax?: number
}

export interface AddSpoolRequest {
  brand: string
  material: string
  colorName: string
  colorHex: string
  initialWeightG: number
  currentWeightG: number
  spoolWeightG?: number
  lowStockThresholdG?: number
  isActive?: boolean
  notes?: string
  density?: number
  diameterTolerance?: number
  extruderMin?: number
  extruderMax?: number
  bedMin?: number
  bedMax?: number
  tagUid?: string
  price?: number
  stockLocation?: string
}

export interface AssignPrinterRequest {
  printerId: string | null
  amsSlot?: number | null
  /** Storage location for the spool currently in the target slot (if any). */
  displacedStockLocation?: string | null
}

export interface SpoolMonthlyStats {
  added: number
  removed: number
}
